framework/test/util/TestUtils.cpp

```diff
@@ -0,0 +1,490 @@
+}
+
+#elif defined(US_PLATFORM_POSIX)
+
+HighPrecisionTimer::HighPrecisionTimer()
+{
+    startTime.tv_nsec = 0;
+    startTime.tv_sec = 0;
+}
+
+void HighPrecisionTimer::Start()
+{
+    clock_gettime(CLOCK_MONOTONIC, &startTime);
+}
+
+long long HighPrecisionTimer::ElapsedMilli()
+{
+    timespec current;
+    clock_gettime(CLOCK_MONOTONIC, &current);
+    return (static_cast<long long>(current.tv_sec) * 1000 + current.tv_nsec / 1000 / 1000) -
+        (static_cast<long long>(startTime.tv_sec) * 1000 + startTime.tv_nsec / 1000 / 1000);
+}
+
+long long HighPrecisionTimer::ElapsedMicro()
+{
+    timespec current;
+    clock_gettime(CLOCK_MONOTONIC, &current);
+    return (static_cast<long long>(current.tv_sec) * 1000 * 1000 + current.tv_nsec / 1000) -
+        (static_cast<long long>(startTime.tv_sec) * 1000 * 1000 + startTime.tv_nsec / 1000);
+}
+
+#elif defined(US_PLATFORM_WINDOWS)
+
+HighPrecisionTimer::HighPrecisionTimer()
+{
+    if (!QueryPerformanceFrequency(&timerFrequency))
+        throw std::runtime_error("QueryPerformanceFrequency() failed");
+}
+
+void HighPrecisionTimer::Start()
+{
+    //DWORD_PTR oldmask = SetThreadAffinityMask(GetCurrentThread(), 0);
+    QueryPerformanceCounter(&startTime);
+    //SetThreadAffinityMask(GetCurrentThread(), oldmask);
+}
+
+long long HighPrecisionTimer::ElapsedMilli()
+{
+    LARGE_INTEGER current;
+    QueryPerformanceCounter(&current);
+    return (current.QuadPart - startTime.QuadPart) / (timerFrequency.QuadPart / 1000);
+}
+
+long long HighPrecisionTimer::ElapsedMicro()
+{
+    LARGE_INTEGER current;
+    QueryPerformanceCounter(&current);
+    return (current.QuadPart - startTime.QuadPart) / (timerFrequency.QuadPart / (1000 * 1000));
+}
+
+#endif
+
+
+Bundle InstallLib(BundleContext frameworkCtx, const std::string& libName)
+{
+  std::vector<Bundle> bundles;
+
+#if defined (US_BUILD_SHARED_LIBS)
+  bundles = frameworkCtx.InstallBundles(LIB_PATH + util::DIR_SEP + US_LIB_PREFIX + libName + US_LIB_EXT);
+#else
+  bundles = frameworkCtx.GetBundles();
+#endif
+
+  for (auto b : bundles)
+  {
+    if (b.GetSymbolicName() == libName) return b;
+  }
+  return {};
+}
+
+void ChangeDirectory(const std::string& destdir)
+{
+  errno = 0;
+  int ret = chdir(destdir.c_str());
+  if (ret != 0)
+  {
+    const std::string msg = "Unable to change directory to " + destdir + ": " + util::GetLastCErrorStr();
+    throw std::runtime_error(msg);
+  }
+}
+
+std::string GetTempDirectory()
+{
+#if defined (US_PLATFORM_WINDOWS)
+  std::wstring temp_dir;
+  wchar_t wcharPath[MAX_PATH];
+  if (GetTempPathW(MAX_PATH, wcharPath))
+  {
+    temp_dir = wcharPath;
+  }
+
+  return std::string(temp_dir.cbegin(), temp_dir.cend());
+#else
+  char* tempdir = getenv("TMPDIR");
+  return std::string(((tempdir == nullptr)?"/tmp":tempdir));
+#endif
+}
+
+static const char validLetters[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+
+// A cross-platform version of the mkstemps function
```

Comment:
Sascha, curious about why we need to be compatible with mkstemp (or mkdtemp), AFAIU, these are only called by our tests and if we drop the compatibility criterion, we could avoid passing "char *" and avoid using strlen() etc. i.e. operate on std::string's instead of (less safe) C-style strings, use std::string methods and throw an exception from this function itself in-case of any error.

Reply:
The answer is a historical one. The originates from the public domain on the Internet. As such it was C code, written to be compatible witk mkstemp and friends by the original author. I only did very small changes originally (many years ago) and used the code as is in another project. It did its job well enough and I thought I would spend my time elsewhere instead of making it prettier for CppMicroServices.

Feel free to send a PR for C++ification or other improvements, but I think there is not much gain unless there is a bug of course.

Reply:
Thanks. Sounds good.